pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -392,8 +390,126 @@ func addDataOutputsToDAG(dag *dependency.DAG, outs []v1alpha2.DataOutput, obj *u
+	}
+	if !ready {
+		return false, nil
+	}
+
+	for i, ct := range acc.Traits {
+		ready, err := r.handleDataInput(ctx, ct.DataInputs, dag, &w.Traits[i])
+		if err != nil {
+			return false, errors.Wrap(err, fmt.Sprintf("handleDataInput for trait (%s/%s) failed", w.Traits[i].GetNamespace(), w.Traits[i].GetName()))
+		}
+		if !ready {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (r *components) handleDataInput(ctx context.Context, ins []v1alpha2.DataInput, dag *dependency.DAG, obj *unstructured.Unstructured) (bool, error) {
 	for _, in := range ins {
-		dag.AddSink(in.ValueFrom.DataOutputName, obj, attaches, in.ToFieldPaths)
+		s, ok := dag.Sources[in.ValueFrom.DataOutputName]
+		if !ok {
+			return false, errors.New("DataOutput name doesn't exist")
+		}
+		val, ready, err := r.checkSourceReady(ctx, s)
+		if err != nil {
+			return false, errors.Wrap(err, "checkSourceReady failed")
+		}
+		if !ready {
+			return false, nil
+		}
+
+		err = fillValue(obj, in.ToFieldPaths, val)
+		if err != nil {
+			return false, errors.Wrap(err, "fillValue failed")
+		}
+	}
+	return true, nil
+}
+
+func fillValue(obj *unstructured.Unstructured, fs []string, val string) error {
+	paved := fieldpath.Pave(obj.Object)
+	for _, fp := range fs {
+		err := paved.SetString(fp, val)
+		if err != nil {
+			return fmt.Errorf("paved.SetString() failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *components) checkSourceReady(ctx context.Context, s *dependency.Source) (string, bool, error) {
+	obj := s.ObjectRef
+	key := types.NamespacedName{
+		Namespace: obj.Namespace,
+		Name:      obj.Name,
+	}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(obj.GroupVersionKind())
+	err := r.client.Get(ctx, key, u)
+	if err != nil {
+		return "", false, fmt.Errorf("failed to get object (%s): %w", key.String(), err)
+	}
+	paved := fieldpath.Pave(u.UnstructuredContent())
+
+	// TODO: Currently only string value supported. Support more types in the future.
+	val, err := paved.GetString(obj.FieldPath)
+	if err != nil {
+		if fieldpath.IsNotFound(err) {
+			return "", false, nil
+		}
+		return "", false, fmt.Errorf("failed to get field value (%s) in object (%s): %w", obj.FieldPath, key.String(), err)
+	}
+
+	ok, err := matchValue(s.Matchers, val, paved)
+	if err != nil {
+		return val, false, err
+	}
+	if !ok {
+		return val, false, nil
+	}
+
+	return val, true, nil
+}
+
+func matchValue(ms []v1alpha2.ConditionRequirement, val string, paved *fieldpath.Paved) (bool, error) {
+	// If no matcher is specified, it is by default to check value not empty.
+	if len(ms) == 0 {
+		return val != "", nil
+	}
+
+	for _, m := range ms {
+		var checkVal string
+		if m.FieldPath != "" {
+			var err error
+			checkVal, err = paved.GetString(m.FieldPath)
+			if err != nil {
+				return false, err
+			}
+		} else {
+			checkVal = val
+		}
+
+		switch m.Operator {
+		case v1alpha2.ConditionEqual:
+			if m.Value != checkVal {
```

Comment:
do we assume it's a primitive type like string or int?

Reply:
Currently only string type is supported. This should fix all the use cases as mentioned in the design doc.
We don't see any use case for int yet.

Reply:
For complex objects, we are going to provide more high level functions like AutoBinding to fill the gap.